tests/test_autodiff.py

```diff
@@ -1,8 +1,83 @@
 import hypothesis
-#import hypothesis.strategies as st
+import hypothesis.strategies as st
 import numpy as np
 import slimdiffy.autodiff as ad
-from function_corpus import get_test_samples, basic_tensor_tests, all_pairs_tests
+from slimdiffy.autodiff import grad, Gradient
+import slimdiffy.pytree as pt
+from function_corpus import (
+    get_test_samples, basic_tensor_tests, all_pairs_tests,
+    TestFunc, tensor_courpus_add, elementwise_strategy
+)
+
+def _to_numpy(x):
+    if isinstance(x, pt.Node):
+        return _to_numpy(x.leaf_value)
+    if isinstance(x, ad.Tracer):
+        equation = x.supervisor.equations[x.idx]
+        print(f"Converting equation type: {type(equation)}")
+        if isinstance(equation, ad.Literal):
+            print(f"Found Literal with value type: {type(equation.value)}")
+            return equation.value
+        if hasattr(equation, 'array'):
+            print(f"Found array with shape: {equation.array.shape}")
+            return equation.array
+        return _to_numpy(equation)
+    if isinstance(x, np.ndarray):
+        return x
+    return np.array(x)
```

Comment:
You shouldn't be handling all these different cases. Just expect the output that you get